File: Leetcode/easy/MajorityElement.scala
// Given an array nums of size n, return the majority element. The majority element is the element that appears more than ⌊n / 2⌋ times.
// Time O(n), Space O(1)

object Solution {
    def majorityElement(nums: Array[Int]): Int = {
        nums.foldLeft((0, 0)) { case ((res, count), num) => {
            val r = if (count == 0) num else res
            val c = if (r == num) count + 1 else count - 1
            (r, c)
        }}._1
    }
}

File: Leetcode/easy/MajorityElement.js
// Given an array nums of size n, return the majority element. The majority element is the element that appears more than ⌊n / 2⌋ times.
// Time O(n), Space O(n)

const majorityElement = function(nums) {
    let count = {}

    for (let n of nums) {
        count[n] ? count[n]++ : count[n] = 1
        if (count[n] > nums.length / 2) return n
    }
}